### ISSUE TITLE
Crash when clicking a calltip away

### ISSUE THREAD

**Issue by tmontes** [jupyter/qtconsole#383]
## Context

Hello all, I've been contributing to the [Mu Editor](https://github.com/mu-editor/mu) -- a beginner friendly Python editor -- which builds its GUI on top of Qt and, among other things, uses `qtconsole` for its Python REPL pane. Before anything else, thanks for sharing `qtconsole`. :)

## The failure

Out of luck I detected [a critical bug](https://github.com/mu-editor/mu/issues/934) that leads to a Mu crash that seems to point to a misbehaviour in `qtconsole` (or, maybe, somewhere else).

In short, when a call tip is visible, clicking with the mouse somewhere within the `qtconsole` widget in an attempt to close the call tip, leads to an exception like:
```
Traceback (most recent call last):
  File ".../qtconsole/call_tip_widget.py", line 89, in hideEvent
    self._cursor_position_changed)
TypeError: 'method' object is not connected
```

The interesting (sad?) part is that such exception leads to a hard exit with PyQt5 but, apparently, not so with PySide2.

## Steps to reproduce

From my testing, with both the Mu Editor and a "plain" `jupyter-qtconsole`, it only seems to fail this way on macOS. Somehow Windows and Linux don't behave exactly the same.

* macOS 10.12.6
* CPython 3.6.8 package from www.python.org installed.

```
$ git clone https://github.com/jupyter/qtconsole/
$ cd qtconsole
$ python3.6 -m venv venv
$ . venv/bin/activate
(venv) $ pip install -e .
(venv) $ pip install pyqt5
(venv) $ jupyter-qtconsole
```

Then, in the GUI `qtconsole`:
```
In [1]: print(
```

At this point, a call tip for `print` pops up. Now, mouse click anywhere on the `qtconsole` widget.

The traceback is printed out and the process terminates with:
```
Traceback (most recent call last):
  File ".../qtconsole/call_tip_widget.py", line 89, in hideEvent
    self._cursor_position_changed)
TypeError: 'method' object is not connected
Abort trap: 6
(venv) $
```

## What I tried

Investigated a bit and found out:

* Running the above with PySide2 instead of PyQt5 produces the same exception and traceback, but doesn't crash the running process. Instead, PySide2 seems to handle it better, printing out `RuntimeError: Failed to disconnect signal cursorPositionChanged().`, after the traceback.
* The failure seems to result from the fact that `hideEvent` in `CallTipWidget` is being called twice -- add a `print` to `showEvent` and `hideEvent` to see that.
* The following platform differences are interesting, not sure if relevant. Whenever the call tip is visible:
  * On Windows and Linux, clicking somewhere "below" the blinking cursor has no effect. Only clicking somewhere "above" does make the call tip go away.
  * On macOS, clicking anywhere leads to the observed failure.

## Eventually related

I found [this Spyder issue](https://github.com/spyder-ide/spyder/issues/10050) showing a similar exception, apparently on macOS, given the file path, which was fixed [with this PR here](https://github.com/jupyter/qtconsole/pull/367). I'm not sure that PR is the "correct" / "best" fix and I wonder if in that case `hideEvent` wasn't being called twice, too. If so, that might be the root cause for the failure.

## Wrap up

Again, thanks for `qtconsole`. I'm not familiar with its design or codebase, much less with PyQt5's. I'll be happy to collaborate in order to see this failure understood and fixed and I'm not sure this is a `qtconsole` issue, or something Qt related. Thanks in advance for your support in this. :)

**Comment by ccordoba12** [comment]
> I'm not sure that PR is the "correct" / "best" fix and I wonder if in that case hideEvent wasn't being called twice, too. If so, that might be the root cause for the failure.

You're welcome to improve things if you think my PR was not the right solution.

**Comment by tmontes** [comment]
@ccordoba12, thanks for your feedback.

I'll be glad to contribute such improvement, if the hypothesis I established is ever confirmed. :)
The observed fact, leading to the crash I'm describing, is that, for some reason:

* The `hideEvent` method, in `CallTipWidget`, is being called twice, leading to the `TypeError` exception I documented and the subsequent crash.

Wrapping the code in `hideEvent` in a `try`/`except TypeError: pass` would mitigate the failure, pretty much like the fix in #367.

What I wonder, and thus the hypothesis I formulated, is whether or not `hideEvent` in `CompletionWidget`, which was patched then, was being called twice too. If that is a fact (I have no idea, at this momen!), then the *root cause* of both issues might be the same and, who knows, lead to other misbehaviours in the future.

With this said, AFAICT, the root cause of this issue seems to be the fact that `hideEvent` in `CallTipWidget` is being called twice. From preliminary observation of the API, it doesn't make sense but, like I said, I'm not familiar with this `qtconsole`'s design/codebase, nor with PyQt5's, so I might be mistaken.

I would love to hear what other, more experienced `qtconsole` users have to say about this.
:)

**Comment by rayosborn** [comment]
I can't claim to be an experienced `qtconsole` user, but I can confirm that this is still an issue, and correctly diagnosed by @tmontes. In my package, it doesn't cause a PyQt5 abort, but I think that is because I have a [customized sys.excepthook function](https://riverbankcomputing.com/pipermail/pyqt/2017-July/039411.html). It is a serious irritation though. Is this issue still under consideration by the qtconsole developers? I have tested the [solution](https://github.com/jupyter/qtconsole/pull/367) implemented by the Spyder project and it does fix the problem with only a few lines of code. I can submit a PR if this hasn't been done before.

**Comment by rayosborn** [comment]
I just had a quick look at closed PRs and notice that the equivalent solution was implemented in #367 for `completion_widget.py`. The same fix needs to be applied to `call_tip_widget.py` as well. I will submit a PR since #367 was approved.

**Comment by ccordoba12** [comment]
Thanks for your help @rayosborn!
